lib/gitsh/readline_history_filter.rb

```diff
@@ -0,0 +1,21 @@
+require 'gitsh/module_delegator'
+
+class ReadlineHistoryFilter < ModuleDelegator
+  def readline(prompt, add_hist = false)
+    module_delegator_target.readline(prompt, add_hist).tap do |input|
+      if add_hist && input && (input.strip.empty? || duplicate?(input))
```

Comment:
I guess we'd still need `add_hist && should_not_have_added_to_history?`, but that still seems more clear.